src/main/kotlin/com/sourcegraph/cody/chat/AgentChatSession.kt

```diff
@@ -173,12 +201,15 @@ private constructor(
   }
 
   @RequiresEdt
-  private fun addMessage(message: ChatMessage) {
-    if (messages.lastOrNull()?.id == message.id) {
-      messages.removeLast()
+  private fun addMessageAtIndex(message: ChatMessage, index: Int) {
+    val messageToUpdate = messages.getOrNull(index)
+    if (messageToUpdate != null) {
+      messages[index] = message
+    } else {
+      messages.add(message)
     }
-    messages.add(message)
-    chatPanel.addOrUpdateMessage(message)
+    chatPanel.addOrUpdateMessage(
+        message, index, shouldAddBlinkingCursor = message.actualMessage().isBlank())
```

Comment:
That have one downside, it will start showing blinking cursor just when we get first blank response from the assistant, right?
If e.g. agent is restarting that could be a few sec.
Maybe it would be better to add it without any conditions later, under:

```kotlin
 chatSession.addMessageAtIndex(
           ChatMessage(
               Speaker.ASSISTANT,
               Source.CHAT,
               text = "",
               displayText = "",
           ),
           chatSession.messages.count())
```
I think here we know we surely need it, no?